### DISCUSSION TITLE
Parser ignores DEPs previously set by AttributeRuler

### DISCUSSION THREAD

**Discussion by pb-jeff-oneill** [explosion/spaCy#12337] [Help: Coding & Implementations] [answered]
My goal is to add some constraints to the tagger/parser so I have a mix of rule-based and model tagging/parsing. For example, suppose I want the word "foo" to always have TAG/DEP of DT/det.

I added an attribute ruler before the tagger to set TAG/DEP for the word "foo".  

The attribute ruler works and the tagger uses the TAG set by the attribute ruler.

The parser, however, ignores the DEP set by the attribute ruler.  Is there a way to have the parser work like the tagger and keep any input DEP attributes?

```
import spacy

def print_tag_dep(nlp, title):
    doc = nlp('foo dog')
    print(title)
    print('Pipeline: ', list(nlp.pipe_names))
    print(f'nlp("foo dog") -> foo (tag: {doc[0].tag_}, dep: {doc[0].dep_}) dog (tag: {doc[1].tag_}, dep: {doc[1].dep_})')
    print()

nlp = spacy.load('en_core_web_sm',
                 exclude=['attribute_ruler', 'lemmatizer', 'ner'])
print_tag_dep(nlp, 'en_core_web_sm')

ruler1 = nlp.add_pipe("attribute_ruler", name='my_ar', before='tagger')
patterns = [[{"ORTH": "foo"}]]
attrs = {"TAG": "DT", "POS": "DET", "DEP":"det"}
ruler1.add(patterns=patterns, attrs=attrs)
print_tag_dep(nlp, 'en_core_web_sm with attribute ruler')

nlp.remove_pipe('parser')
print_tag_dep(nlp, 'en_core_web_sm with attribute ruler and without parser')
```

```
en_core_web_sm
Pipeline:  ['tok2vec', 'tagger', 'parser']
nlp("foo dog") -> foo (tag: NN, dep: compound) dog (tag: NN, dep: ROOT)

en_core_web_sm with attribute ruler
Pipeline:  ['tok2vec', 'my_ar', 'tagger', 'parser']
nlp("foo dog") -> foo (tag: DT, dep: compound) dog (tag: NN, dep: ROOT)

en_core_web_sm with attribute ruler and without parser
Pipeline:  ['tok2vec', 'my_ar', 'tagger']
nlp("foo dog") -> foo (tag: DT, dep: det) dog (tag: NN, dep: )
```
I'm on Mac with Python 3.7 and Spacy 3.3.2.

Edited to more clearly show the precise problem.

**Comment by adrianeboyd** [comment, marked as answer]
No, there isn't. Some components can work around partial annotation to some degree (most of these have an `overwrite` setting so you can decide whether to preserve existing annotation), but the parser isn't one of them. It will always clobber all the existing annotation. The internal parsing algorithm just isn't intended to handle this case (in exchange it's very efficient).

Also be aware that that even though the tagger doesn't clobber existing annotation with `overwrite=False`, it also doesn't use the existing annotation in any way to inform its predictions, so the existing annotation won't affect the tags of the surrounding words.

**Comment by pb-jeff-oneill** [comment]
Thanks, this is very helpful!
